feat: reintroduce raw-strings

mrdgo opened the pull request:
[Discussion with tree-sitter devs](https://github.com/tree-sitter/tree-sitter/discussions/3878#discussioncomment-11139978)

This will break the current `nvim-treesitter` config. They have to update to also use `scanner.c`. People using that config might want to use "our" way to install ts-nu.

After merge, it might be necessary to run `:TSUpdate nu` at least once. I don't know about treesitter's caching, so let's just hope that nobody's setup breaks.

With that scanner, we can also create custom logic for unquoted strings.

mrdgo commented:
@blindFS
Feel free to try this yourself! The code stands, it passes tests and correctly highlights in neovim-nightly. Please follow the link in the description (discussion with tree-sitter-devs) and see the current show-stopper. As soon as I start the exact same setup within neovim 0.10.2/0.10.1/0.9.5, it crashes without any error message. The only thing that I found out is a null-pointer dereference - which doesn't make sense, because the scanner works standalone and in neovim-nightly. I am deeply puzzled and don't really have an idea how to procceed

siph commented:
fwiw this is working for me (`3c50ee4186153dd768141c9d0a2fcc1fdcd47b03`) in `v0.10.2`:
![Screenshot_20241120_144941](https://github.com/user-attachments/assets/5a7a8937-8730-4e8d-89b8-9ad389c039a6)

![Screenshot_20241120_145028](https://github.com/user-attachments/assets/9e1dbf18-082a-4993-a2a8-1dbe8397dd7a)

I'm using nix to build the grammar and neovim.

blindFS commented:
I also have no problem with your changes made on-top of current main branch:

<img width="644" alt="image" src="https://github.com/user-attachments/assets/6b469171-5793-418c-874e-36d2b9518fd5">

<img width="397" alt="image" src="https://github.com/user-attachments/assets/85772cde-4fa0-4eb3-a164-47312700c65f">

My neovim is installed from nix-darwin for macos.

mrdgo commented on tree-sitter.json:
Is this file even required? Just noticed that I pushed it.

clason replied:
It's newly created by the latest tree-sitter CLI. (Please make sure you always use a released version of the CLI -- 0.24.4 right now -- not a prerelease commit since downstream consumers will not have adapted to the breaking changes that will come with tree-sitter 0.25.)

mrdgo commented on src/scanner.c:
```diff
@@ -0,0 +1,172 @@
+    lexer->log(lexer, "Nu Scanner: level [%i], emitted [%b]\n", s->level,
+               s->emitted_content);
+
+    if (valid_symbols[RAW_STRING_BEGIN] && s->level == 0) {
```
done

clason commented on .github/neovim/init.lua:
```diff
@@ -0,0 +1,37 @@
+-- Bootstrap lazy.nvim
+local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
```
You don't need `lazy`; just clone and set up nvim-treesitter directly.

You might also want to consider using the `main` branch for tests; that is simpler to set up and has much better performance (but still WIP), see https://github.com/tree-sitter/tree-sitter/blob/master/.github/workflows/nvim_ts.yml

Make sure you include a `generate` step!

(You can call `./scripts/install-parsers.lua nu` to only install your parser.)

Check out https://github.com/tree-sitter/workflows as well.

mrdgo replied:
I am not happy with the pipeline, you reinforced this gut feeling. I will take a look, thanks!

clason commented on .github/workflows/main.yml:
```diff
@@ -30,3 +30,9 @@ jobs:
+    uses: ./.github/workflows/nvim-tests.yml
+    with:
+      neovim_versions: |
+        [ "nightly", "v0.10.2", "v0.10.1", "v0.9.5" ]
```
nvim-treesitter no longer supports v0.9 and earlier.

mrdgo replied:
thanks!

blindFS commented on src/scanner.c:
```diff
@@ -0,0 +1,172 @@
+    return false;
+}
+
+bool scan_raw_string_end(TSLexer *lexer, Scanner *s) {
```
I can pass all your test cases with the following code:
```c
static bool scan_raw_string_end(TSLexer *lexer, Scanner *s) {
    lexer->log(lexer, "END\n");
    adv;
    for (unsigned i = 0; i < s->level; i++) {
        adv;
    }
    s->level = 0;
    return true;
}
```

<img width="366" alt="image" src="https://github.com/user-attachments/assets/706ed699-74b6-4114-9681-18fee6513bde">

Better to put the `raw-strings.nu` in `expr` ?

blindFS replied:
Let me try to explain my understanding of the logic here:

1. in `scan_raw_string_content` we already make sure the ending level is correct,
   - if it's not, for example, at the 2nd  `'` in `r##' hello'# '##`, it won't stop the while loop, it will advance further until the last `'` is reached, then mark_end there.
2. then in `scan_raw_string_end`, we simply advance `s->level` + 1 (for the `'`) times.

clason commented:
And in general, I would strongly recommend to make conceptually unrelated changes in separate PRs; that makes it much easier to review and merge preliminary steps.

(You can cherry-pick commits from PRs that are required to be merged first, as long as you clearly label them as such.)

mrdgo commented:
Incorporated all reviews, removed the pipeline changes that neither provide insights nor prevent anything. Maybe pipeline is a separate issue that should be addressed at some point.

fdncred commented:
Ping me if/when we're ready to land this and I'll push the button.

mrdgo commented:
Thanks to @blindFS for the thorough review! Is there someone else who would want to review? Otherwise, I'd be ready to land, again

mrdgo commented:
Aaah, wait a second

blindFS commented on src/scanner.c:
```diff
@@ -0,0 +1,163 @@
+        if (level == s->level) {
+            lexer->log(lexer, "Detected end\n" );
+            return true;
+        } else if (level == s->level && len == 0) {
```
No need for this branch now.

mrdgo replied:
true, removed

blindFS commented on src/scanner.c:
```diff
@@ -0,0 +1,163 @@
+    // scan for r#' r##' or more #
+    skip_whitespace(lexer);
+
+    if (lexer->lookahead == 'r') {
```
This check is already done.

mrdgo replied:
This also makes `skip_whitespace` obsolete, thanks!

blindFS commented on queries/nu/highlights.scm:
```diff
@@ -90,6 +90,10 @@ file_path: (val_string) @variable.parameter
 (escaped_interpolated_content) @string
 (expr_interpolated ["(" ")"] @variable.parameter)
 
+(raw_string_begin) @string
```
We probably don't need this, `val_string` will get it properly highlighted.

mrdgo replied:
True, removed

fdncred commented:
looks like we're closer, just have a conflict now.

fdncred commented:
Are we ready?

blindFS commented:
> Are we ready?

I think so.

mrdgo commented:
@clason does nvim-ts' lockfile interfere, even when I override the `install_info`?

clason commented:
Yes, the lockfile takes precedence. (Nvim-ts is not a general purpose installer; you can shoehorn additional parsers but it's not designed for replacing tracked parsers.)

mrdgo commented:
> Yes, the lockfile takes precedence. (Nvim-ts is not a general purpose installer; you can shoehorn additional parsers but it's not designed for replacing tracked parsers.)

Would you then pretty-please be so kind :innocent:

clason commented:
We have automation for that
